### PR TITLE
Refactor for pull strategy + pull strategy

### PR DESCRIPTION
**Pull request by KostyaSha**
There is no AMI terminology in docker. It was copy-paste from other plugins.
Called new variables to be more generic in case some other plugin decided to copy-paster in future.

Implements #79

### CODE REVIEW

**Comment by KostyaSha** [comment]
Closes #79

File: docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerImagePullStrategy.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package com.nirima.jenkins.plugins.docker;
+
+/**
+ * @author Kanstantsin Shautsou
+ */
+public enum DockerImagePullStrategy {
+
+    PULL_ALWAYS("Pull all images every time") {
```
Comment (lanwen) [inline docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerImagePullStrategy.java]: how can we localize it?

File: docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java [hunk 2]
```diff
@@ -252,16 +279,51 @@ public static String runContainer(DockerTemplateBase dockerTemplateBase,
...
+            LOGGER.info("Pulling image '{}' since one was not found.  This may take awhile...", imageName);
+            //Identifier amiId = Identifier.fromCompoundString(ami);
+            try (InputStream imageStream = getClient().pullImageCmd(imageName).exec()) {
+                int streamValue = 0;
+                while (streamValue != -1) {
+                    streamValue = imageStream.read();
+                }
+            }
+
+            LOGGER.info("Finished pulling image '{}'", imageName);
```
Comment (lanwen) [inline docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java]: if we don't pull image, why we write that we finished?

File: docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerImagePullStrategy.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        public boolean pullIfNotExists(String imageName) {
+            return true;
+        }
+
+        @Override
+        public boolean pullIfExists(String imageName) {
+            return imageName.endsWith(":latest");
+        }
+    },
+    PULL_NEVER("Never pull") {
```
Comment (lanwen) [inline docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerImagePullStrategy.java]: think use case in javadoc will be useful

Comment (KostyaSha) [inline docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerImagePullStrategy.java]: Use case it up to user, just implement ability to not pull any images, i.e. somebody on mobile on dial-up

File: docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerImagePullStrategy.java (hunk 1 above)
Comment (KostyaSha) [inline docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerImagePullStrategy.java]: Probably, https://github.com/jenkinsci/jenkins/blob/9c443c8d5bafd63fce574f6d0cf400cd8fe1f124/core/src/main/java/hudson/model/BallColor.java#L57
But if don't want waste time on localisation :)

File: docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerImagePullStrategy.java (hunk 3 above)
Comment (lanwen) [inline docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerImagePullStrategy.java]: who run slave via mobile internet? :)

Comment (KostyaSha) [inline docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerImagePullStrategy.java]: or just in high cost vm. Do you have any objections for it? ;)

File: docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java (hunk 2 above)
Comment (KostyaSha) [inline docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java]: We are pulling, this block is only for pulling image

Comment (lanwen) [inline docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java]: miss end of "if" statement

**Comment by lanwen** [comment]
for code :+1: , but you should think about localization and tests. Its not a waste of time

**Comment by KostyaSha** [comment]
@lanwen i think functionality is more critical and i'm cycling between different usage blockers. I'm ok to do translations when code will be stable. 
I aslo thinking about uses ReposTag object instead of String, but i fear changed in docker-java and will keep as is for now.

**Comment by KostyaSha** [comment]
@magnayn never comment single commits, use `changes` tab instead

File: docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java [hunk 4]
```diff
@@ -96,16 +100,46 @@ public DockerCloud(String name,
...
     }
 
     public int getConnectTimeout() {
         return connectTimeout;
     }
 
+    /**
+     * @deprecated use {@link #getContainerCap()}
+     */
+    @Deprecated
```
Comment (KostyaSha) [inline docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java]: restored this method to have ABI/API compatibility

File: docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java [hunk 5]
```diff
@@ -252,16 +293,51 @@ public static String runContainer(DockerTemplateBase dockerTemplateBase,
...
+                    streamValue = imageStream.read();
+                }
+            }
+
+            LOGGER.info("Finished pulling image '{}'", imageName);
+        }
+    }
 
     private DockerSlave provisionWithWait(DockerTemplate dockerTemplate) throws IOException, Descriptor.FormException {
+        pullImage(dockerTemplate);
```
Comment (KostyaSha) [inline docker-plugin/src/main/java/com/nirima/jenkins/plugins/docker/DockerCloud.java]: moved `pullImage` into this block

**Comment by KostyaSha** [comment]
Verified manually plus @DanielHouston promised to test it on Monday.
Merging, but @lanwen feel free to review.

**Pull request merged**